backend/store/etcd/watchers.go

```diff
@@ -123,3 +125,52 @@ func GetResourceWatcher(ctx context.Context, client *clientv3.Client, key string
 
 	return ch
 }
+
+// GetResourceV3Watcher returns a channel that emits WatchEventResourceV3
+// structs notifying that a corev3.Resource has been created, deleted or
+// updated.
+func GetResourceV3Watcher(ctx context.Context, client *clientv3.Client, key string) <-chan store.WatchEventResourceV3 {
+	w := Watch(ctx, client, key, true)
+	c := make(chan store.WatchEventResourceV3, 1)
+
+	go func() {
+		defer close(c)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+
+			case event := <-w.Result():
+				logger := logger.WithField("key", event.Key)
+
+				switch action := event.Type; {
+				case action == store.WatchCreate, action == store.WatchDelete, action == store.WatchUpdate:
+					wrapper := &wrap.Wrapper{}
+					if err := wrapper.Unmarshal(event.Object); err != nil {
+						logger.WithError(err).Error("unable to unmarshal resource from key")
+						continue
+					}
+
+					component, err := wrapper.Unwrap()
+					if err != nil {
+						logger.WithError(err).Error("unable to unwrap wrapper")
+						continue
+					}
+
+					c <- store.WatchEventResourceV3{
+						Action:   action,
+						Resource: component,
+					}
+
+				case action == store.WatchError, action == store.WatchUnknown:
```

Comment:
Maybe a logger statement here too?

Reply:
In the case for `WatchError`, is `event.Err` carrying the actualy error or not? I wasn't 100% sure after a cursory look at the underlying watcher code...

I've updated the PR with logging, but in doubt I did not include `.WithError(event.Err)`. Let me know if I should correct that.